pkg/cache.go

```diff
@@ -0,0 +1,57 @@
+package pkg
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type MetricsCache struct {
+	cacheMutex   *sync.Mutex
+	entries      map[string]CacheEntry
+	ttlInSeconds int
+}
+
+func NewMetricsCache(ttlInSeconds int) *MetricsCache {
+	return &MetricsCache{
+		cacheMutex:   &sync.Mutex{},
+		entries:      map[string]CacheEntry{},
+		ttlInSeconds: ttlInSeconds,
+	}
+}
+
+func getMetricHash(metric prometheus.Metric) string {
+	checksum := sha256.Sum256([]byte(metric.Desc().String()))
```

Comment:
I am afraid we need to change the hash function a bit.
This only takes into account the static labels added during creation of the `metric.Desc`. Dynamic labels like `aws_region` are not respected when the hash is generated leading to those overwriting themselves.

You can add the following test that replicates that behaviour:

```
func TestSameMetricWithDifferentLabelsDontOverwrite(t *testing.T) {
	cache := NewMetricsCache(1)
	desc := prometheus.NewDesc("test", "multimetric", []string{"aws_region"}, nil)

	metricEast1 := prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 1, "us-east-1")
	metricWest1 := prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 1, "us-west-1")
	metricEast2 := prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 2, "us-east-1")
	cache.AddMetric(metricEast1)
	cache.AddMetric(metricWest1) // should *not* overwrite metricEast1
	cache.AddMetric(metricEast2) // should overwrite metricEast1

	assert.Len(t, cache.GetAllMetrics(), 2)
	assert.NotContains(t, cache.GetAllMetrics(), metricEast1)
	assert.Contains(t, cache.GetAllMetrics(), metricWest1)
	assert.Contains(t, cache.GetAllMetrics(), metricEast2)
}
```

Reply:
Good catch! thank's for testing @mrWinston and @bergmannf 
Perhaps it's possible to use the Write method for exporting a hashable representation?

Reply:
Hm, for me GetLabel gives me these:
```
metric.Write(&dto)
fmt.Println(dto.GetLabel())
output;
[name:"aws_region" value:"us-east-1" ]
```

Reply:
I pushed a commit, that should fix the hashing, please take a look. 
The string generated for hashing now looks like this:

```
Desc{fqName: "test", help: "multimetric", constLabels: {}, variableLabels: [aws_region]},aws_region,us-east-1
```

Reply:
I'll resolve this thread, since the test you provided looks good.